### PR TITLE
Fix: noon and midnight adjustment

### LINKED ISSUE #902: Adaptive lighting always sets light to 1%
# Home Assistant Adaptive Lighting Issue Template

### Before submitting a bug report, please follow these troubleshooting steps:

Please confirm that you have completed the following steps:

- [x] I have updated to the [latest Adaptive Lighting version](https://github.com/basnijholt/adaptive-lighting/releases) available in [HACS](https://hacs.xyz/). DONE
- [x] I have reviewed the [Troubleshooting Section](https://github.com/basnijholt/adaptive-lighting#troubleshooting) in the [README](https://github.com/basnijholt/adaptive-lighting#readme). DONE
- [x] (If using Zigbee2MQTT) I have read the [Zigbee2MQTT troubleshooting guide](https://github.com/basnijholt/adaptive-lighting#zigbee2mqtt) in the [README](https://github.com/basnijholt/adaptive-lighting#readme). NA
- [x] I have checked the [V2 Roadmap](https://github.com/basnijholt/adaptive-lighting/discussions/291) and [open issues](https://github.com/basnijholt/adaptive-lighting/issues) to ensure my issue isn't a duplicate. DONE

### Required information for bug reports:

Please include the following information in your issue.

*Issues missing this information may not be addressed.*

1.  **Debug logs** captured while the issue occurred. [See here for instructions on enabling debug logging](https://github.com/basnijholt/adaptive-lighting#troubleshooting):

```
Turn off
2024-01-13 18:14:59.771 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_off('['light.mbed_bulb_left', 'light.mbed_bulb_right']', transition=None)' event with context.id='01HM20HGXQ966NBJVR023BJ9W3'
2024-01-13 18:14:59.951 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'on' → 'off' event for 'light.mbed_bulb_right' with context.id='01HM20HGXQ966NBJVR023BJ9W3'
2024-01-13 18:14:59.987 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'on' → 'off' event for 'light.mbed_bulb_left' with context.id='01HM20HGXQ966NBJVR023BJ9W3'

Turn On
2024-01-13 18:15:12.747 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] (0) _service_interceptor_turn_on_handler: call.context.id='01HM20HXKA3ZE1DXN1D7Y3DNST', is_skipped_hash='False'
2024-01-13 18:15:12.748 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] (1) _service_interceptor_turn_on_handler: call='<ServiceCall light.turn_on (c:01HM20HXKA3ZE1DXN1D7Y3DNST): entity_id=['light.mbed_side'], params=>', service_data='{'entity_id': ['light.mbed_side'], 'params': {}}'
2024-01-13 18:15:12.749 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] No switch found for entity_id='light.mbed_side', skipped='['light.mbed_side']'
2024-01-13 18:15:12.750 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] (2) _service_interceptor_turn_on_handler: switch_to_eids='{}', skipped='['light.mbed_side']'
2024-01-13 18:15:12.752 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] (0) _service_interceptor_turn_on_handler: call.context.id='01HM20HXKA3ZE1DXN1D7Y3DNST', is_skipped_hash='False'
2024-01-13 18:15:12.752 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] (1) _service_interceptor_turn_on_handler: call='<ServiceCall light.turn_on (c:01HM20HXKA3ZE1DXN1D7Y3DNST): entity_id=['light.mbed_bulb_left', 'light.mbed_bulb_right'], params=>', service_data='{'entity_id': ['light.mbed_bulb_left', 'light.mbed_bulb_right'], 'params': {}}'
2024-01-13 18:15:12.756 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Switch is off or light is already on for entity_id='light.mbed_bulb_left', skipped='[]' (is_on='True', is_state='False', manual_control='False', switch._intercept='False')
2024-01-13 18:15:12.757 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Switch is off or light is already on for entity_id='light.mbed_bulb_right', skipped='['light.mbed_bulb_left']' (is_on='True', is_state='False', manual_control='False', switch._intercept='False')
2024-01-13 18:15:12.757 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] (2) _service_interceptor_turn_on_handler: switch_to_eids='{}', skipped='['light.mbed_bulb_left', 'light.mbed_bulb_right']'
2024-01-13 18:15:12.759 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.mbed_bulb_left', 'light.mbed_bulb_right']')' event with context.id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.925 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.mbed_bulb_right' 'state_changed' event: '{'min_color_temp_kelvin': 2202, 'max_color_temp_kelvin': 4000, 'min_mireds': 250, 'max_mireds': 454, 'effect_list': ['colorloop'], 'supported_color_modes': [<ColorMode.COLOR_TEMP: 'color_temp'>, <ColorMode.XY: 'xy'>], 'color_mode': <ColorMode.COLOR_TEMP: 'color_temp'>, 'brightness': 254, 'color_temp_kelvin': 2202, 'color_temp': 454, 'hs_color': (29.79, 84.553), 'rgb_color': (255, 146, 39), 'xy_color': (0.579, 0.388), 'effect': None, 'off_with_transition': False, 'off_brightness': None, 'friendly_name': 'MBed Bulb Right Light', 'supported_features': <LightEntityFeature.EFFECT|FLASH|TRANSITION: 44>}' with context.id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.926 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'off' → 'on' event for 'light.mbed_bulb_right' with context.id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.926 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] is_proactively_adapting_context='False', context_id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.926 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] just_turned_off: State change 'off' → 'on' triggered by 'light.turn_on'
2024-01-13 18:15:12.927 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] is_proactively_adapting_context='False', context_id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.930 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: '_update_attrs_and_maybe_adapt_lights' called with context.id='01HM20HXRZ:al:JVBG:lght:19' lights: '['light.mbed_bulb_right']', transition: '0.0', force: 'True'
2024-01-13 18:15:12.934 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: filtered_lights: '['light.mbed_bulb_right']'
2024-01-13 18:15:12.934 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] is_proactively_adapting_context='False', context_id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.935 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: Calling _adapt_light from _update_attrs_and_maybe_adapt_lights: 'light.mbed_bulb_right' with transition 0.0 and context.id=01HM20HXRZ:al:JVBG:lght:19
2024-01-13 18:15:12.962 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: Setting color_temp of light light.mbed_bulb_right
2024-01-13 18:15:12.963 DEBUG (MainThread) [custom_components.adaptive_lighting.adaptation_utils] Preparing adaptation data for light.mbed_bulb_right with service data {'entity_id': 'light.mbed_bulb_right', 'brightness': 3, 'color_temp_kelvin': 2202}
2024-01-13 18:15:12.963 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: execute_cancellable_adaptation_calls with data: AdaptationData(entity_id=light.mbed_bulb_right, context_id=01HM20HXRZ:al:JVBG:lght:19, sleep_time=0.0, force=True, max_length=1, which=both, initial_sleep=False)
2024-01-13 18:15:12.984 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.mbed_bulb_right', 'brightness': 3, 'color_temp_kelvin': 2202} with context.id='01HM20HXRZ:al:JVBG:lght:19'
2024-01-13 18:15:12.992 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.mbed_bulb_left' 'state_changed' event: '{'min_color_temp_kelvin': 2202, 'max_color_temp_kelvin': 4000, 'min_mireds': 250, 'max_mireds': 454, 'effect_list': ['colorloop'], 'supported_color_modes': [<ColorMode.COLOR_TEMP: 'color_temp'>, <ColorMode.XY: 'xy'>], 'color_mode': <ColorMode.COLOR_TEMP: 'color_temp'>, 'brightness': 254, 'color_temp_kelvin': 2202, 'color_temp': 454, 'hs_color': (29.79, 84.553), 'rgb_color': (255, 146, 39), 'xy_color': (0.579, 0.388), 'effect': None, 'off_with_transition': False, 'off_brightness': None, 'friendly_name': 'Mbed Bulb Left Light', 'supported_features': <LightEntityFeature.EFFECT|FLASH|TRANSITION: 44>}' with context.id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.993 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'off' → 'on' event for 'light.mbed_bulb_left' with context.id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.993 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] is_proactively_adapting_context='False', context_id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.993 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] just_turned_off: State change 'off' → 'on' triggered by 'light.turn_on'
2024-01-13 18:15:12.993 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] is_proactively_adapting_context='False', context_id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:12.994 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: '_update_attrs_and_maybe_adapt_lights' called with context.id='01HM20HXV2:al:JVBG:lght:1A' lights: '['light.mbed_bulb_left']', transition: '0.0', force: 'True'
2024-01-13 18:15:12.997 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: filtered_lights: '['light.mbed_bulb_left']'
2024-01-13 18:15:13.000 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] is_proactively_adapting_context='False', context_id='01HM20HXKA3ZE1DXN1D7Y3DNST'
2024-01-13 18:15:13.000 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: Calling _adapt_light from _update_attrs_and_maybe_adapt_lights: 'light.mbed_bulb_left' with transition 0.0 and context.id=01HM20HXV2:al:JVBG:lght:1A
2024-01-13 18:15:13.011 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.mbed_bulb_right']')' event with context.id='01HM20HXRZ:al:JVBG:lght:19'
2024-01-13 18:15:13.012 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] (0) _service_interceptor_turn_on_handler: call.context.id='01HM20HXRZ:al:JVBG:lght:19', is_skipped_hash='False'
2024-01-13 18:15:13.016 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: Setting color_temp of light light.mbed_bulb_left
2024-01-13 18:15:13.016 DEBUG (MainThread) [custom_components.adaptive_lighting.adaptation_utils] Preparing adaptation data for light.mbed_bulb_left with service data {'entity_id': 'light.mbed_bulb_left', 'brightness': 3, 'color_temp_kelvin': 2202}
2024-01-13 18:15:13.017 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: execute_cancellable_adaptation_calls with data: AdaptationData(entity_id=light.mbed_bulb_left, context_id=01HM20HXV2:al:JVBG:lght:1A, sleep_time=0.0, force=True, max_length=1, which=both, initial_sleep=False)
2024-01-13 18:15:13.026 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] MBed Side Bulbs: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.mbed_bulb_left', 'brightness': 3, 'color_temp_kelvin': 2202} with context.id='01HM20HXV2:al:JVBG:lght:1A'
2024-01-13 18:15:13.033 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.mbed_bulb_left']')' event with context.id='01HM20HXV2:al:JVBG:lght:1A'
2024-01-13 18:15:13.034 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] (0) _service_interceptor_turn_on_handler: call.context.id='01HM20HXV2:al:JVBG:lght:1A', is_skipped_hash='False'
2024-01-13 18:15:13.269 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.mbed_bulb_right' 'state_changed' event: '{'min_color_temp_kelvin': 2202, 'max_color_temp_kelvin': 4000, 'min_mireds': 250, 'max_mireds': 454, 'effect_list': ['colorloop'], 'supported_color_modes': [<ColorMode.COLOR_TEMP: 'color_temp'>, <ColorMode.XY: 'xy'>], 'color_mode': <ColorMode.COLOR_TEMP: 'color_temp'>, 'brightness': 3, 'color_temp_kelvin': 2202, 'color_temp': 454, 'hs_color': (29.79, 84.553), 'rgb_color': (255, 146, 39), 'xy_color': (0.579, 0.388), 'effect': None, 'off_with_transition': False, 'off_brightness': None, 'friendly_name': 'MBed Bulb Right Light', 'supported_features': <LightEntityFeature.EFFECT|FLASH|TRANSITION: 44>}' with context.id='01HM20HXRZ:al:JVBG:lght:19'
2024-01-13 18:15:13.274 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] AdaptiveLightingManager: New adapt '<state light.mbed_bulb_right=on; min_color_temp_kelvin=2202, max_color_temp_kelvin=4000, min_mireds=250, max_mireds=454, effect_list=['colorloop'], supported_color_modes=[<ColorMode.COLOR_TEMP: 'color_temp'>, <ColorMode.XY: 'xy'>], color_mode=color_temp, brightness=3, color_temp_kelvin=2202, color_temp=454, hs_color=(29.79, 84.553), rgb_color=(255, 146, 39), xy_color=(0.579, 0.388), effect=None, off_with_transition=False, off_brightness=None, friendly_name=MBed Bulb Right Light, supported_features=44 @ 2024-01-13T18:15:12.916318+00:00>' found for light.mbed_bulb_right
2024-01-13 18:15:13.275 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] No transition in last adapt for light light.mbed_bulb_right, not starting timer.
2024-01-13 18:15:13.315 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.mbed_bulb_left' 'state_changed' event: '{'min_color_temp_kelvin': 2202, 'max_color_temp_kelvin': 4000, 'min_mireds': 250, 'max_mireds': 454, 'effect_list': ['colorloop'], 'supported_color_modes': [<ColorMode.COLOR_TEMP: 'color_temp'>, <ColorMode.XY: 'xy'>], 'color_mode': <ColorMode.COLOR_TEMP: 'color_temp'>, 'brightness': 3, 'color_temp_kelvin': 2202, 'color_temp': 454, 'hs_color': (29.79, 84.553), 'rgb_color': (255, 146, 39), 'xy_color': (0.579, 0.388), 'effect': None, 'off_with_transition': False, 'off_brightness': None, 'friendly_name': 'Mbed Bulb Left Light', 'supported_features': <LightEntityFeature.EFFECT|FLASH|TRANSITION: 44>}' with context.id='01HM20HXV2:al:JVBG:lght:1A'
2024-01-13 18:15:13.319 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] AdaptiveLightingManager: New adapt '<state light.mbed_bulb_left=on; min_color_temp_kelvin=2202, max_color_temp_kelvin=4000, min_mireds=250, max_mireds=454, effect_list=['colorloop'], supported_color_modes=[<ColorMode.COLOR_TEMP: 'color_temp'>, <ColorMode.XY: 'xy'>], color_mode=color_temp, brightness=3, color_temp_kelvin=2202, color_temp=454, hs_color=(29.79, 84.553), rgb_color=(255, 146, 39), xy_color=(0.579, 0.388), effect=None, off_with_transition=False, off_brightness=None, friendly_name=Mbed Bulb Left Light, supported_features=44 @ 2024-01-13T18:15:12.970162+00:00>' found for light.mbed_bulb_left
2024-01-13 18:15:13.319 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] No transition in last adapt for light light.mbed_bulb_left, not starting timer.

```

2.  [Your Adaptive Lighting configuration](https://github.com/basnijholt/adaptive-lighting#gear-configuration):

I have set AL using the Lovelace and do not know how to get the configuration.  Looked at README and googled but no joy.  Please point me to how to do this and I will add information.  However, the configuration has been working for monthes and has not been changed.

4.  Describe the bug and how to reproduce it:

When I turn the light on which used AL it always goes to 1%, 

6. Steps to reproduce the behavior:

turn light to full then turn off and turn on

### PR DESCRIPTION
**Pull request by AndreasBackx**
Might resolve #902 and other issues related to incorrectly adjusting noon and midnight times.

I live in London and today is the day where solar midnight is very close to actual midnight. It's around 23:59. What happened this evening is that my lights suddenly went from their normal percentage to 1%. (It wasn't sleep mode.) My guess (and @Viproz's in #902) is also that the custom midnight calculation in `noon_and_midnight` is incorrect. When the custom logic is triggered by for example setting `min_sunset_time`, `prev_and_next_events` returned the event types sunset, sunrise when it should have returned sunset, solar_midnight.

I debugged and it's clear that when this occurs, the custom code does not fix the day properly:

Inside of `SunEvents.prev_and_next_events` it calculates yesterday's, today's, and tomorrow's events. If we log those, we get this when having not changed any values that would trigger the custom logic:

```
0. event_name='solar_midnight'
0. event_time_formatted=datetime.datetime(2024, 4, 16, 23, 59, 42)
1. event_name='sunrise'
1. event_time_formatted=datetime.datetime(2024, 4, 17, 4, 59, 32, 723716)
2. event_name='solar_noon'
2. event_time_formatted=datetime.datetime(2024, 4, 17, 11, 59, 48)
3. event_name='sunset'
3. event_time_formatted=datetime.datetime(2024, 4, 17, 19, 0, 56, 58566)
4. event_name='solar_midnight'
4. event_time_formatted=datetime.datetime(2024, 4, 17, 23, 59, 29)
5. event_name='sunrise'
5. event_time_formatted=datetime.datetime(2024, 4, 18, 4, 57, 25, 655122)
6. event_name='solar_noon'
6. event_time_formatted=datetime.datetime(2024, 4, 18, 11, 59, 35)
7. event_name='sunset'
7. event_time_formatted=datetime.datetime(2024, 4, 18, 19, 2, 36, 688279)
8. event_name='solar_midnight'
8. event_time_formatted=datetime.datetime(2024, 4, 18, 23, 59, 16)
9. event_name='sunrise'
9. event_time_formatted=datetime.datetime(2024, 4, 19, 4, 55, 19, 432385)
10. event_name='solar_noon'
10. event_time_formatted=datetime.datetime(2024, 4, 19, 11, 59, 21)
11. event_name='sunset'
11. event_time_formatted=datetime.datetime(2024, 4, 19, 19, 4, 17, 265827)
```

You can see that it always follows the same pattern cyclically: `solar_midnight`, `sunrise`, `solar_noon`, `sunset`, and so on. When changing for example `min_sunset_time` and `max_sunset_time` even though they don't affect the eventually calculated sunset time, it will give the following incorrect result:

```
0. event_name='solar_midnight'
0. event_time_formatted=datetime.datetime(2024, 4, 17, 0, 0, 14, 391141)
1. event_name='sunrise'
1. event_time_formatted=datetime.datetime(2024, 4, 17, 4, 59, 32, 723716)
2. event_name='solar_noon'
2. event_time_formatted=datetime.datetime(2024, 4, 17, 12, 0, 14, 391141)
3. event_name='sunset'
3. event_time_formatted=datetime.datetime(2024, 4, 17, 19, 0, 56, 58566)
4. event_name='solar_midnight'
4. event_time_formatted=datetime.datetime(2024, 4, 18, 0, 0, 1, 171700)
5. event_name='sunrise'
5. event_time_formatted=datetime.datetime(2024, 4, 18, 4, 57, 25, 655122)
6. event_name='solar_noon'
6. event_time_formatted=datetime.datetime(2024, 4, 18, 12, 0, 1, 171700)
7. event_name='sunset'
7. event_time_formatted=datetime.datetime(2024, 4, 18, 19, 2, 36, 688279)
8. event_name='sunrise'
8. event_time_formatted=datetime.datetime(2024, 4, 19, 4, 55, 19, 432385)
9. event_name='solar_noon'
9. event_time_formatted=datetime.datetime(2024, 4, 19, 11, 59, 48, 349106)
10. event_name='sunset'
10. event_time_formatted=datetime.datetime(2024, 4, 19, 19, 4, 17, 265827)
11. event_name='solar_midnight'
11. event_time_formatted=datetime.datetime(2024, 4, 19, 23, 59, 48, 349106)
```

You can see that it incorrectly added a day to the last solar midnight leading to `sunset` -> `sunrise` instead of `sunset` -> `solar_midnight` -> `sunrise`.

I've fixed it so we do not need to any special calculations to figure out whether to add/remove a day. I am calculating the offset between the "adjusted offset" and "real sunset" and apply the same delta to midnight. (The same for sunrise and noon.) This leads to the above example returning the correct first example.

When we change the `sunset_max_time` to something like 4PM, we can also see that the difference is applied correctly:

```
0. event_name='solar_midnight'
0. event_time_formatted=datetime.datetime(2024, 4, 16, 20, 58, 45, 941434)
1. event_name='sunrise'
1. event_time_formatted=datetime.datetime(2024, 4, 17, 4, 59, 32, 723716)
2. event_name='solar_noon'
2. event_time_formatted=datetime.datetime(2024, 4, 17, 11, 59, 48)
3. event_name='sunset'
3. event_time_formatted=datetime.datetime(2024, 4, 17, 16, 0)
4. event_name='solar_midnight'
4. event_time_formatted=datetime.datetime(2024, 4, 17, 20, 56, 52, 311721)
5. event_name='sunrise'
5. event_time_formatted=datetime.datetime(2024, 4, 18, 4, 57, 25, 655122)
6. event_name='solar_noon'
6. event_time_formatted=datetime.datetime(2024, 4, 18, 11, 59, 35)
7. event_name='sunset'
7. event_time_formatted=datetime.datetime(2024, 4, 18, 16, 0)
8. event_name='solar_midnight'
8. event_time_formatted=datetime.datetime(2024, 4, 18, 20, 54, 58, 734173)
9. event_name='sunrise'
9. event_time_formatted=datetime.datetime(2024, 4, 19, 4, 55, 19, 432385)
10. event_name='solar_noon'
10. event_time_formatted=datetime.datetime(2024, 4, 19, 11, 59, 21)
11. event_name='sunset'
11. event_time_formatted=datetime.datetime(2024, 4, 19, 16, 0)
```

I primarily tested this with sunset and midnight, though it should work the same for sunrise and noon.

### CODE REVIEW

**Comment by AndreasBackx** [comment]
I'm not entirely sure why `test_light_settings` in `test_switch.py` started to fail. It's a bit hard to follow the code there as I'm not familar with Home Assistant's code nor the testing code here. @basnijholt, if you have some time you might know what the issue is. Though no rush, thank you!

**Comment by AndreasBackx** [comment]
@basnijholt would it be possible to have a look here? I observed the same issue today/yesterday as today solar noon/midnight slightly differs again: https://www.timeanddate.com/sun/uk/london.

**Comment by basnijholt** [comment]
Thanks for the reminder!

All the tests seem to fail and I personally don't have time to debug it myself. 

If you manage to get the tests to pass I'll happily merge!

**Comment by AndreasBackx** [comment]
@basnijholt 

I verified that 2 tests have pre-existing failures, those are:
- `test_switch.py::test_proactive_adaptation_with_separate_commands`
- `test_light_switch_in_specific_area`
	- I was able to fix this one easily, see the trivial change in `mock_area_registry`.

Though I did break the test `test_switch.py::test_light_settings`. However after spending quite some time trying to debug this, I cannot figure out why `assert_expected_color_temp` is called. It seems to assure that the color temperature does not change, though I'm not sure why it shouldn't change. If you could help share your experience with the codebase, we could hopefully figure out why it's failing so I can fix it.

At the same time, I fixed the `Dockerfile` to build again because the paths were incorrect and it would say the tests folder from `CMD` could not be found. Additionally, I've made it so `docker build` handles caching better and so every single change does not trigger reinstalling all Python dependencies.

File: scripts/setup-dependencies [hunk 1]
```diff
@@ -2,7 +2,7 @@
 set -ex
 cd "$(dirname "$0")/.."
 
-pip install -r core/requirements.txt
+pip install -r core/requirements.txt --root-user-action
```
Comment (AndreasBackx) [inline scripts/setup-dependencies:5]:
```suggestion
pip install -r core/requirements.txt --root-user-action=ignore
```

This needs to be added to all of them. Seems like this option saw changes in different pip versions.
